src/identity.rs

```diff
@@ -0,0 +1,844 @@
+    pub fn mapped_by(&self, node: &NodeIdentity, role: &str, group: &str) -> bool {
+        for (pod_id, mapped_node) in self.mapping.iter() {
+            if node == mapped_node && pod_id.role() == role && pod_id.group() == group {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Given `pods` return all that are not mapped.
+    pub fn missing(&self, pods: &[PodIdentity]) -> Vec<PodIdentity> {
+        let mut result = vec![];
+        for p in pods {
+            if !self.mapping.contains_key(p) {
+                result.push(p.clone())
+            }
+        }
+        result
+    }
+
+    #[cfg(test)]
+    pub fn new(map: Vec<(PodIdentity, NodeIdentity)>) -> Self {
+        let mut result = BTreeMap::new();
+        for (p, n) in map {
+            result.insert(p.clone(), n.clone());
+        }
+        PodToNodeMapping { mapping: result }
+    }
+}
+
+/// A pod identity generator that can be implemented by the operators.
+///
+/// Implementation of this trait are responsible for:
+/// - generating all pod identities expected by the service by implementing `PodIdentityFactory::as_slice`
+/// - map pods to their identity values by implementing `PodIdentityFactory::try_mapping`
+pub trait PodIdentityFactory {
+    /// A slice with all pod identities that should exist for a given service.
+    fn as_slice(&self) -> &[PodIdentity];
+    /// Returns a PodToNodeMapping for the given pods or an error if any pod could not be mapped.
+    fn try_from(&self, pods: &[Pod]) -> Result<Vec<PodIdentity>, Error>;
+}
+
+/// An implementation of [`PodIdentityFactory`] where id's are incremented across all roles and groups
+/// contained in `eligible_nodes`.
+///
+/// This factory requires pods to be labeled with all `REQUIRED_LABELS` and with one additional "id label"
+/// that can vary from operator to operator.
+///
+/// See `generate_ids` for details.
+pub struct LabeledPodIdentityFactory {
+    app: String,
+    instance: String,
+    id_label_name: String,
+    slice: Vec<PodIdentity>,
+}
+
+impl LabeledPodIdentityFactory {
+    /// Build a new instance of this factory.
+    ///
+    /// See `Self::generate_ids` for implemtation details.
+    ///
+    /// Clippy complains that this function is unused but this is not true. It is used in tests
+    /// and it is meant to be called by the operators. It's part of this module's API.
+    /// Also clippy complains about the `generate_ids` which is called here.
+    /// For this reason, this function is marked as "dead code".
+    ///
+    /// # Arguments
+    /// - `app` : Application name.
+    /// - `instance` : Application name.
+    /// - `eligible_nodes` : Eligible nodes (and pod replicas) grouped by role and group.
+    /// - `id_label_name` : Name of the pod's id label used to store the `id` field of `PodIdentity`
+    /// - `start` : The initial value when generating the `id` fields of pod identities.
+    #[allow(dead_code)]
```

Comment:
You do not need that `#[allow(dead_code)]`. Clippy is not complaining. Its a pub method.
Maybe clean your workspace and try again.

Reply:
Done.